Cherry-picked SHOW command PRs (#3080)
neo4j/docs-operations: modules/ROOT/pages/procedures/show-procedures.adoc

:description: Information about the `SHOW PROCEDURES` command.
= Show procedures

You can use `SHOW PROCEDURES` to list all available procedures.

> For general information about the `SHOW` command, see the link:{neo4j-docs-base-uri}/cypher-manual/clauses/show/[Cypher Manual -> SHOW].

Reviewer: ```suggestion
For general information about the `SHOW` command, see the link:{neo4j-docs-base-uri}/cypher-manual/5/clauses/show/[Cypher Manual -> SHOW].
```

====

|===

[NOTE]
====
When using the `RETURN` clause, the `YIELD` clause is mandatory and must not be omitted.
====

[[return-columns]]
== Return columns

`SHOW PROCEDURES` returns following columns:

.Show procedures output
[options="header", cols="3m,4a,2m,1a"]
|===
| Column | Description | Type | Default output

| name
| The name of the procedure.
| STRING
| {check-mark}

| description
| The procedure description.
| STRING
| {check-mark}

| mode
| The procedure mode, for example `READ` or `WRITE`.
| STRING
| {check-mark}

| worksOnSystem
| Whether the procedure can be run on the `system` database or not.
| BOOLEAN
| {check-mark}

| signature
| The signature of the procedure.
| STRING
|

| argumentDescription
| List of the arguments for the procedure, as map of strings and booleans with `name`, `type`, `default`, `isDeprecated`, and `description`.
| LIST<MAP>
|

| returnDescription
| List of the returned values for the procedure, as map of strings and booleans with `name`, `type`, `isDeprecated`, and `description`.
| LIST<MAP>
|

| admin
| `true` if this procedure is an admin procedure.
| BOOLEAN
|

| rolesExecution
| List of roles permitted to execute this procedure.

> Is `null` without the link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/dbms-administration/dbms-role-management-privileges/#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.

Reviewer 2: is that the link we want to use? feels like we rarely use the `neo4j.com/docs` type of links....

> Is `null` without the link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/dbms-administration/dbms-role-management-privileges/#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.

Reviewer: ```suggestion
Is `null` without the xref:authentication-authorization/dbms-administration/dbms-role-management-privileges.adoc#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.
```

| LIST<STRING>
|

| rolesBoostedExecution
| List of roles permitted to use boosted mode when executing this procedure.

> Is `null` without the link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/dbms-administration/dbms-role-management-privileges/#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.

Reviewer 2: is that the link we want to use? feels like we rarely use the `neo4j.com/docs` type of links....

> Is `null` without the link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/dbms-administration/dbms-role-management-privileges/#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.

Reviewer: ```suggestion
Is `null` without the xref:authentication-authorization/dbms-administration/dbms-role-management-privileges.adoc#_grant_privilege_to_show_roles[`SHOW ROLE`] privilege.
```

| LIST<STRING>
|

| isDeprecated

> | Whether the procedure is deprecated.

Reviewer 2: missing the introduced in label of the current page
```suggestion
| abel:new[Introduced in 5.9] Whether the procedure is deprecated.
```

| BOOLEAN
|

| deprecatedBy

> | The replacement procedure to use in case of deprecation; otherwise `null`.

Reviewer 2: same here, missing introduced in
```suggestion
| label:new[Introduced in 5.21] The replacement procedure to use in case of deprecation; otherwise `null`.
```

.Show all procedures and return all columns
[source, cypher]
----
SHOW PROCEDURES YIELD *
----

.Show all procedures and return specific columns only
[source, cypher]
----
SHOW PROCEDURES YIELD name, admin
----

.Filter `SHOW PROCEDURES` using `WHERE`
[source, cypher]
----
SHOW PROCEDURES YIELD name, worksOnSystem
WHERE worksOnSystem = TRUE
RETURN name
----

The listed procedures can also be filtered by whether a user can execute them.
This filtering is only available through the `EXECUTABLE` clause and not through the `WHERE` clause.
This is due to using the user's privileges instead of filtering on the available output columns.

There are two options for using the `EXECUTABLE` clause.
The first option is to filter for the current user:

.Show procedures executable by the current user
[source, cypher, role=test-result-skip]
----
SHOW PROCEDURES EXECUTABLE BY CURRENT USER YIELD *
----

The second option for using the `EXECUTABLE` clause is to filter the list to only contain procedures executable by a specific user.
The following example shows the procedures available to the user `jake`, who has been granted the `EXECUTE PROCEDURE dbms.*` privilege by the `admin` of the database.

> For more information about `DBMS EXECUTE` privilege administration, see the xref:authentication-authorization/dbms-administration/dbms-execute-privileges.adoc[The `DBMS EXECUTE` privileges].

Reviewer 2: the old page had the tag
```suggestion
For more information about `DBMS EXECUTE` privilege administration, see the xref:authentication-authorization/dbms-administration/access-control-dbms-administration-execute.adoc[The `DBMS EXECUTE` privileges].
```
so I guess that might be the Cypher 5 version of the tag?

Reviewer: The correct link is as Jens suggested:

For more information about `DBMS EXECUTE` privilege administration, see the xref:authentication-authorization/dbms-administration/dbms-execute-privileges.adoc[The `DBMS EXECUTE` privileges].

Reviewer 2: I guess this was missed being updated when that changed then 🤷
